test/logger-errors.js

```diff
@@ -0,0 +1,385 @@
+      return true
+    })
+    .query(() => { return true })
+    .reply(200, 'Ingester response')
+
+  logger.on('error', (err) => {
+    t.type(err, Error, 'Expected to be an Error')
+    t.match(err, {
+      name: 'Error'
+    , message: 'Invalid log level.  Using the default instead.'
+    , meta: {
+        got: 'NEIGH'
+      , expected: constants.LOG_LEVELS
+      , used: logger.level
+      }
+    }, 'Expected Error is correct')
+  })
+  logger.log('log line', {level: 'NEIGH'})
+})
+
+test('.log() ignores empty or null messages', (t) => {
+  t.plan(6)
+
+  const logger = new Logger(apiKey, createOptions())
+  const expected = [null, undefined, '']
+
+  logger.on('warn', (obj) => {
+    t.match(obj, {
+      message: 'Log statement was empty.  Ignored'
+    }, `Got warning for ${obj.statement}`)
+    t.ok(expected.includes(obj.statement), 'statement property is correct')
+  })
+
+  logger.log(null)
+  logger.log(undefined)
+  logger.log('')
+})
+
+test('removeMetaProperty emits \'warn\' if property is not found', (t) => {
+  t.plan(1)
+  const logger = new Logger(apiKey)
+
+  logger.on('removeMetaProperty', t.fail)
+  logger.on('warn', (obj) => {
+    t.deepEqual(obj, {
+      message: 'Property is not an existing meta property.  Cannot remove.'
+    , key: 'NOPE'
+    }, 'Got expected warning')
+  })
+  logger.removeMetaProperty('NOPE')
+})
+
+test('HTTP timeout will emit Error and continue to retry', (t) => {
+  const delay = 1000 // Set this low since nock will ultimately wait to timeout
+  const logger = new Logger(apiKey, createOptions({
+    baseBackoffMs: 100
+  , maxBackoffMs: 500
+  , timeout: delay
+  }))
+  let attempts = 0
+
+  t.on('end', async () => {
+    nock.cleanAll()
+  })
+
+  // Fail 3 times, then succeed. FYI, the axios agent treats a timeout on connection
+  // the same as a timeout on response body (connection accepted; no reply)
+  nock(logger.url)
+    .post('/', () => {
+      t.equal(
+        logger[Symbol.for('isLoggingBackedOff')]
+      , false
+      , 'Logger is not backed off prior to the first failure'
+      )
+      return true
+    })
+    .query(() => { return true })
+    .delayConnection(delay + 1)
+    .reply(200, 'Will Not Happen')
+    .post('/', () => {
+      t.equal(
+        logger[Symbol.for('isLoggingBackedOff')]
+      , true
+      , 'Logger is backed off'
+      )
+      return true
+    })
+    .query(() => { return true })
+    .delayConnection(delay + 1)
+    .reply(200, 'Will Not Happen')
+    .post('/', () => { return true })
+    .query(() => { return true })
+    .delayConnection(delay + 1)
+    .reply(200, 'Will Not Happen')
+    .post('/', () => { return true })
+    .query(() => { return true })
+    .reply(200, 'Success')
+
+  logger.on('error', (err) => {
+    t.type(err, Error, 'Error type is emitted')
+    t.match(err, {
+      message: 'An error occured while sending logs to the cloud.'
+    , meta: {
+        actual: `timeout of ${delay}ms exceeded`
```

Comment:
deepEqual

Reply:
`deepEqual` must match properties exactly, so to use it on an `Error` means you have to have things like the `name` in there.   Testing that deeply on core objects is kinda unnecessary, and it makes for bulkier tests, so typically you'll see `match` with just a few property tests for `message` and `code`.  Here, I wanted to deeply test `meta` since it's my addition.

Here, all I have to do is add `name`, so I see no reason to `deepEqual` it here.  Most of these are just cut and paste from places where I wasn't testing deeply, so thanks for looking closely.

Reply:
Added `name` for `deepEqual`